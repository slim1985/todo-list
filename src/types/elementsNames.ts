export enum ElementsNames {
    STATUS_SELECT = 'status-select',
    TITLE_TEXT_AREA = 'title-textarea',
    DESCRIPTION_TEXT_AREA = 'description-textarea',
}
