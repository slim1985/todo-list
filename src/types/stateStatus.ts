export enum StateStatus {
    LOADING = 'loading',
    IDLE = 'idle',
    FAILED = 'failed',
}
