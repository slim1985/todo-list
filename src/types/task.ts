export interface Task {
    id: string;
    title: string;
    description: string;
    state: TaskStates;
}

export enum TaskStates {
    new = 'new',
    active = 'active',
    completed = 'completed',
}
