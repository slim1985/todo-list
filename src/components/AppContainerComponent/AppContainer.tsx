import { useTasks } from '../../hooks/useTasks';
import { TaskContainer } from '../../components/TaskContainerComponent/TaskContainer';
import { TaskPanel } from '../../components/TaskPanelComponent/TaskPanel';
import { User } from '../../types/user';

export interface AppContainerProps {
    user: User | null;
    signOut: () => void;
}

export function AppContainer({
    user,
    signOut,
}: AppContainerProps): JSX.Element {
    const [
        taskList,
        stateStatus,
        showTaskForm,
        selectedTask,
        hideTaskForm,
        createTask,
        updateTask,
        deleteTask,
        openTask,
        clearTaskList,
    ] = useTasks();

    return (
        <div>
            {!showTaskForm && (
                <TaskPanel
                    user={user}
                    signOut={signOut}
                    clearTaskList={clearTaskList}
                    openTask={openTask}
                />
            )}
            <TaskContainer
                taskList={taskList}
                stateStatus={stateStatus}
                showTaskForm={showTaskForm}
                selectedTask={selectedTask}
                hideTaskForm={hideTaskForm}
                createTask={createTask}
                updateTask={updateTask}
                deleteTask={deleteTask}
                openTask={openTask}
            />
        </div>
    );
}
