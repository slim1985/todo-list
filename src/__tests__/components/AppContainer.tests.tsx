import { render } from '@testing-library/react';
import { AppContainer } from '../../components/AppContainerComponent/AppContainer';
import { Task, TaskStates } from '../../types/task';
import { StateStatus } from '../../types/stateStatus';
import { User } from '../../types/user';

const user: User = {
    id: '1',
    displayName: 'Test User',
    email: 'user@user.com',
};

const taskList: Task[] = [
    {
        id: '1',
        title: 'Task 1',
        description: 'Task 1 description',
        status: TaskStates.NEW,
    },
    {
        id: '2',
        title: 'Task 2',
        description: 'Task 2 description',
        status: TaskStates.ACTIVE,
    },
    {
        id: '3',
        title: 'Task 3',
        description: 'Task 3 description',
        status: TaskStates.COMPLETED,
    },
];

const signOut = jest.fn();

jest.mock('../../store/taskSlice', () => ({
    getTasksAsync: jest.fn(),
    createTaskAsync: jest.fn(),
    updateTaskAsync: jest.fn(),
    deleteTaskAsync: jest.fn(),
    selectAllTasks: jest.fn(),
    clearTasks: jest.fn(),
}));

jest.mock('../../store/useTaskSelector', () => ({
    useTaskSelector: jest.fn(),
}));

jest.mock('../../store/useTaskDispatch', () => ({
    useTaskDispatch: jest.fn(),
}));

jest.mock('../../store/store', () => ({
    RootState: jest.fn(),
    store: {
        getState: jest.fn(),
    },
}));

jest.mock('../../hooks/useTasks', () => ({
    useTasks: (): useTasksType => getUseTasksMocks(false),
}));

describe('AppContainer', () => {
    it('AppContainer renders correctly without showTaskForm', () => {
        // Arrange.
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        jest.spyOn(require('../../hooks/useTasks'), 'useTasks').mockReturnValue(
            getUseTasksMocks(false),
        );

        // Act.
        const { asFragment } = render(
            <AppContainer user={user} signOut={signOut} />,
        );

        // Assert.
        expect(asFragment()).toMatchSnapshot();
    });

    it('AppContainer renders correctly with showTaskForm', () => {
        // Arrange.
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        jest.spyOn(require('../../hooks/useTasks'), 'useTasks').mockReturnValue(
            getUseTasksMocks(true),
        );

        // Act.
        const { asFragment } = render(
            <AppContainer user={user} signOut={signOut} />,
        );

        // Assert.
        expect(asFragment()).toMatchSnapshot();
    });
});

function getUseTasksMocks(showTaskForm: boolean): useTasksType {
    return [
        taskList,
        StateStatus.IDLE,
        showTaskForm,
        showTaskForm ? taskList[0] : null,
        jest.fn(),
        jest.fn(),
        jest.fn(),
        jest.fn(),
        jest.fn(),
        jest.fn(),
    ];
}

type useTasksType = [
    taskList: Task[],
    stateStatus: StateStatus,
    showTaskForm: boolean,
    selectedTask: Task | null,
    hideTaskForm: () => void,
    createTask: (
        title: string,
        description: string,
        status: TaskStates,
    ) => void,
    updateTask: (
        id: string,
        title: string,
        description: string,
        status: TaskStates,
    ) => void,
    deleteTask: (taskId: string) => void,
    openTask: (taskId: string | null) => void,
    clearTaskList: () => void,
];
