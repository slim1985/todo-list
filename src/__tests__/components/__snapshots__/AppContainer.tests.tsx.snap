// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContainer AppContainer renders correctly with showTaskForm 1`] = `
<DocumentFragment>
  <div>
    <div
      class="flex justify-center"
    >
      <div>
        <div
          class="flex flex-wrap justify-center mt-3"
        >
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 1
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Task 1 description
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-green-700"
              >
                New
              </span>
            </p>
          </div>
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 2
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Task 2 description
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-blue-700"
              >
                Active
              </span>
            </p>
          </div>
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 3
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Task 3 description
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-gray-700"
              >
                Completed
              </span>
            </p>
          </div>
        </div>
      </div>
      <div
        class="flex justify-center items-center z-10 fixed h-full w-full overflow-hidden bg-black/[.5]"
      >
        <div
          class="flex flex-col flex-nowrap w-5/6 h-5/6 bg-white bg-gray-300 rounded-md retina:rounded-3xl"
        >
          <select
            class="text-3xl retina:text-8xl m-1 retina:m-3"
            data-testid="status-select"
            name="status"
          >
            <option
              value="0"
            >
              New
            </option>
            <option
              value="1"
            >
              Active
            </option>
            <option
              value="2"
            >
              Completed
            </option>
          </select>
          <textarea
            class="h-24 retina:h-64 border-solid border-2 retina:border-4 border-yellow-200 rounded retina:rounded-2xl m-1 retina:m-3 p-1 retina:p-3 retina:text-6xl"
            data-testid="title-textarea"
            name="title"
          >
            Task 1
          </textarea>
          <textarea
            class="h-full border-solid border-2 retina:border-4 border-yellow-200 rounded retina:rounded-2xl m-1 retina:m-3 p-1 retina:p-3 retina:text-6xl"
            data-testid="description-textarea"
            name="description"
          >
            Task 1 description
          </textarea>
          <div
            class="flex justify-center my-5 retina:my-5 mx-3 retina:mx-8 space-x-7"
          >
            <button
              class="size-20 retina:size-60 p-1 retina:p-3 bg-green-300 rounded-md retina:rounded-3xl retina:text-6xl"
            >
              Save
            </button>
            <button
              class="size-20 retina:size-60 p-1 retina:p-3 bg-yellow-300 rounded-md retina:rounded-3xl retina:text-6xl"
            >
              Cancel
            </button>
            <button
              class="size-20 retina:size-60 p-1 retina:p-3 bg-red-300 rounded-md retina:rounded-3xl retina:text-6xl"
            >
              Delete
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AppContainer AppContainer renders correctly without showTaskForm 1`] = `
<DocumentFragment>
  <div>
    <div
      class="flex justify-start sticky top-0 space-x-3 bg-white solid border-gray-400 border-b-2 retina:border-b-4"
    >
      <div
        class="flex w-2/4 justify-start items-center float-left"
      >
        <div
          class="ml-2 retina:ml-6"
        >
          <p
            class="font-semibold retina:text-5xl"
          >
            Test User
          </p>
          <p
            class="retina:mt-3 text-sm retina:text-5xl"
          >
            user@user.com
          </p>
          <button
            class="mt-2 retina:mt-5 retina:text-5xl"
          >
            Sign Out
          </button>
        </div>
      </div>
      <div
        class="flex w-2/4 justify-end"
      >
        <button
          class="size-20 retina:size-52 my-2 retina:my-4 mr-2 retina:mr-5 p-1 retina:text-5xl bg-gray-300 rounded-md retina:rounded-3xl"
        >
          Create
        </button>
      </div>
    </div>
    <div
      class="flex justify-center"
    >
      <div>
        <div
          class="flex flex-wrap justify-center mt-3"
        >
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 1
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Task 1 description
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-green-700"
              >
                New
              </span>
            </p>
          </div>
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 2
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Task 2 description
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-blue-700"
              >
                Active
              </span>
            </p>
          </div>
          <div
            class="flex flex-col flex-nowrap w-64 md:w-96 retina:w-4/12 h-48 md:h-64 retina:h-192 m-1 retina:m-4 p-2 retina:p-8 items-stretch justify-left cursor-pointer border-solid border-2 retina:border-4 border-gray-400 rounded-lg retina:rounded-3xl"
          >
            <h2
              class="w-full font-bold retina:text-6xl"
            >
              Task 3
            </h2>
            <p
              class="w-full h-screen py-2 retina:py-8 overflow-hidden text-ellipsis text-sm retina:text-6xl"
            >
              Task 3 description
            </p>
            <p
              class="w-full font-bold retina:text-6xl"
            >
              Task state: 
              <span
                class="text-gray-700"
              >
                Completed
              </span>
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
